plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/file/FileMetastoreTableOperations.java

```diff
@@ -76,7 +77,7 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
             metastore.replaceTable(database, tableName, table, privileges);
         }
         catch (RuntimeException e) {
-            throw new CommitFailedException(e, "Failed to commit transaction to FileHiveMetastore");
+            throw new CommitStateUnknownException(e);
```

Comment:
thanks for fixing this!

wondering, can there be still a situation while handling `RuntimeException` that we need to `CommitFailedException` to make sure its retried and cleaned up the Iceberg Library?
Or was it introduced to handle any specific case?

Reply:
We'd need to handle transient and non-transient errors differently, which we don't do today and is hard to do since it depends on the underlying implementation. In general, permanent (non-transient) errors are not expected to happen frequently, so I don't think we need to worry about it.

If this causes a problem for someone, we can look at the specific situation and figure out how to handle it.

Reply:
There are a few tests:

- `TestIcebergOrcConnectorTest.testInsertRowConcurrently` 
- `TestIcebergConnectorSmokeTest.testDeleteRowsConcurrently`

which failed due to concurrent modifications exceptions.


I added a special handling in the `FileMetastoreTableOperations`

```
            if (e instanceof TrinoException trinoException &&
                    trinoException.getErrorCode() == HIVE_METASTORE_ERROR.toErrorCode() &&
                    Objects.equals(trinoException.getMessage(), "Cannot update Iceberg table: supplied previous location does not match current location")) {
                // CommitFailedException is handled as a special case in the Iceberg library. This commit will automatically retry
                throw new CommitFailedException(e, "Failed to replace table due to concurrent updates: %s.%s", database, tableName);
            }
```

@electrum  is this what you were hinting at in  https://github.com/trinodb/trino/issues/14104#issuecomment-1244755074

> We can handle it by checking for the ` supplied previous location does not match current location` exception message.